### PR TITLE
Feature/form helper refactor

### PR DESCRIPTION
**Pull request by katenorquay**

### CODE REVIEW

File: src/form_helper.js [hunk 1]
```diff
@@ -98,37 +102,33 @@ export default class FormHelper {
...
 
-        for (var elementName in this.formHelperConfig[countryCode].elements) {
-          if (this.formHelperConfig[countryCode].elements.hasOwnProperty(elementName)) {
-            const element = this.formHelperConfig[countryCode].elements[elementName];
-
-            if(element && !document.body.contains(element)){
-              this._log(`Element ${elementName} is not in the DOM`)
-              return false
-            }
+        for (var elementName in formConfig.elements) {
```
Comment (nigelramsay) [inline src/form_helper.js]: could you apply the `#find` pattern here?

File: src/form_helper.js [hunk 2]
```diff
@@ -98,37 +102,33 @@ export default class FormHelper {
...
+    const doesntContainElement = element => !document.body.contains(element)
+
+    if( doesntContainElement(this.formHelperConfig.countryElement)){
       this._log("Country Element is not in the DOM")
       return false
     }
 
-    // TODO can we aggregate the elements to check into a single array or map?
-
     const countryCodes = ['nz', 'au']
```
Comment (nigelramsay) [inline src/form_helper.js]: how about defining `country_codes` as a class `const` as it's used in more than one place.

File: src/form_helper.js [hunk 3]
```diff
@@ -154,52 +154,43 @@ export default class FormHelper {
...
 
-        if(!preserveValues){
-          this._clearElementValues("au")
-          this._clearElementValues("nz")
-        }
+    this._setActiveCountry(activeCountry)
+    if(!preserveValues) {
+      const countryCodes = ["au", "nz"]
+      const isInactiveCountry = countryCode => countryCode != activeCountry
+      countryCodes.map((countryCode) => {
```
Comment (nigelramsay) [inline src/form_helper.js]: how about something like:

```
countryCodes.filter(isInactiveCountry).each(this._clearElementValues)
```

or something like this.

File: src/form_helper.js [hunk 4]
```diff
@@ -232,10 +223,8 @@ export default class FormHelper {
     let elements = this.formHelperConfig.au.elements
 
     if(elements.address_line_1_and_2){
-      const combined = [
-        metaData.address_line_1, metaData.address_line_2
-      ].filter(function(a){return a != null}).join(", ")
-
+      const addressisNotNull = array => array != null
```
Comment (nigelramsay) [inline src/form_helper.js]: perhaps remove double-negative `addressIsPresent`

File: src/form_helper.js [hunk 5]
```diff
@@ -256,36 +245,41 @@ export default class FormHelper {
   }
 
   _setElementValue(element, value, elementName){
-    if (element) {
-      element.value = value;
+    if (!element) {
+      this.logErrorMessage(element, value, elementName)
```
Comment (nigelramsay) [inline src/form_helper.js]: hmmm - does this work? the `logErrorMessage` function appears to take `0` parameters.

File: src/form_helper.js [hunk 6]
```diff
@@ -256,36 +245,41 @@ export default class FormHelper {
...
+   }
 
-      var event = document.createEvent('HTMLEvents');
-      event.initEvent('change', true, false);
-      element.dispatchEvent(event);
+   else if (element.options) {
+      const find = (f, x) => Array.prototype.find.call(x, f)
```
Comment (nigelramsay) [inline src/form_helper.js]: why redeclare the `Array.find` function?

File: src/form_helper.js [hunk 7]
```diff
@@ -256,36 +245,41 @@ export default class FormHelper {
...
                        + '\nValue: ' + value;
+   if (window.console) {
+     console.warn(errorMessage);
+   }
+  }
 
-    if (window.console) {
-      console.warn(errorMessage);
-    }
+  createEvent(element) {
```
Comment (nigelramsay) [inline src/form_helper.js]: I would probably not extract this out into a _public function_, as it's fairly specific to the `setElementValue` function. Let's discuss.

File: src/form_helper.js (hunk 2 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 6 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 7 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 5 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 4 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 3 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

File: src/form_helper.js (hunk 1 above)
Comment (katenorquay) [inline src/form_helper.js]: 👍

**Pull request merged**